### ISSUE TITLE
False Positive `useless_attribute` with `#[expect(redundant_imports)]`

### ISSUE THREAD

**Issue by LambdaAlpha** [rust-lang/rust-clippy#15316]
### Summary

When a macro definition is exported by `pub(crate) use` and then used later in the same file but not used in other files, `cargo check` generates `redundant_imports` warning. When `#[expect(redundant_imports)]` is added to the re-exported item, `cargo check` generates no warning bug `cargo clippy` generates `useless_attribute` warning.

### Lint Name

useless_attribute

### Reproducer

I tried this code:

```rust
macro_rules! empty {
    () => {};
}

#[expect(redundant_imports)]
pub(crate) use empty;

empty!();
```

I saw this happen:

```
error: useless lint attribute
 --> src\lib.rs:5:1
  |
5 | #[expect(redundant_imports)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: if you just forgot a `!`, use: `#![expect(redundant_imports)]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_attribute
  = note: `#[deny(clippy::useless_attribute)]` on by default

error: could not compile `lib` (lib) due to 1 previous error
```

I expected to see this happen: no warning

### Version

```text
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-pc-windows-msvc
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by y21** [comment]
Fix should be pretty simple: just need to add the `redundant_imports` lint to https://github.com/rust-lang/rust-clippy/blob/fd7076bd004a457c84a897be56dfb006b07ce26c/clippy_lints/src/attrs/useless_attribute.rs#L28-L39

Would probably also be worth going through the [list of all allow-by-default rustc lints](https://doc.rust-lang.org/beta/rustc/lints/listing/allowed-by-default.html#redundant-imports) and look for more ones that possibly lint `use` items and aren't included there

**Comment by LambdaAlpha** [comment]
I can reproduce this bug using nightly clippy (`cargo +nightly clippy`). Is this fix already available in rustup nightly channel? @y21 

`cargo +nightly clippy --version`:
clippy 0.1.90 (a7a1618e6c 2025-07-22)

**Comment by y21** [comment]
>  Is this fix already available in rustup nightly channel?

No, commits in this repo are only upstreamed to rust-lang/rust every two weeks (from which the component is built), so they aren't always immediately there on the next nightly and could take up to 14 days. That said, from a quick look the last sync happened 13d ago so the next one likely happens in the upcoming days and the fix should be in one of the next nightlies.

If you do want to check if this is really fixed for your project, you could clone this repo and run `cargo dev lint /path/to/your/project`, which runs the very latest clippy.
